lib/Sema/TypeCheckType.cpp

```diff
@@ -1943,7 +1938,34 @@ Type ResolveTypeRequest::evaluate(Evaluator &evaluator,
   if (validateAutoClosureAttributeUse(ctx.Diags, TyR, result, options))
     return ErrorType::get(ctx);
 
-  return result;
+  // Diagnose an attempt to use a placeholder at the top level.
+  if (result->getCanonicalType()->is<PlaceholderType>() &&
+      resolution->getOptions().contains(TypeResolutionFlags::Direct)) {
+    if (!resolution->getOptions().contains(TypeResolutionFlags::SilenceErrors))
+      ctx.Diags.diagnose(loc, diag::top_level_placeholder_type);
+
+    TyR->setInvalid();
+    return ErrorType::get(ctx);
+  }
+
+  // Now that top-level placeholders have been diagnosed, replace them according
+  // to the user-specified handler.
+  return result.get().transform([&](Type ty) {
+    if (auto *oldTy = ty->getAs<PlaceholderType>()) {
+      auto originator = oldTy->getOriginator();
+      if (auto *repr = originator.dyn_cast<PlaceholderTypeRepr *>()) {
+        // Fill in the placeholder if there's an appropriate handler.
+        if (const auto handlerFn = resolution->getPlaceholderHandler())
```

Comment:
Checking `getPlaceholderHandler()` first looks more efficient.
```swift
  if (const auto handlerFn = resolution->getPlaceholderHandler()) {
    result = result.get().transform.transform([&](Type ty) {
       ...
    });
  }
  return result;
```

Reply:
Ah yeah, this should be stable over a given resolution. Good call!